# Regression Assignment:

In [74]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [75]:
dataset = pd.read_csv("insurance_prediction_dataset.csv")

In [76]:
dataset

age     sex     bmi  children smoker      charges
0      19  female  27.900         0    yes  16884.92400
1      18    male  33.770         1     no   1725.55230
2      28    male  33.000         3     no   4449.46200
3      33    male  22.705         0     no  21984.47061
4      32    male  28.880         0     no   3866.85520
...   ...     ...     ...       ...    ...          ...
1333   50    male  30.970         3     no  10600.54830
1334   18  female  31.920         0     no   2205.98080
1335   18  female  36.850         0     no   1629.83350
1336   21  female  25.800         0     no   2007.94500
1337   61  female  29.070         0    yes  29141.36030

[1338 rows x 6 columns]

In [77]:
dataset.shape

(1338, 6)

In [78]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 62.8+ KB


In [79]:
dataset = pd.get_dummies(dataset, dtype=int, drop_first= True)

In [80]:
dataset

age     bmi  children      charges  sex_male  smoker_yes
0      19  27.900         0  16884.92400         0           1
1      18  33.770         1   1725.55230         1           0
2      28  33.000         3   4449.46200         1           0
3      33  22.705         0  21984.47061         1           0
4      32  28.880         0   3866.85520         1           0
...   ...     ...       ...          ...       ...         ...
1333   50  30.970         3  10600.54830         1           0
1334   18  31.920         0   2205.98080         0           0
1335   18  36.850         0   1629.83350         0           0
1336   21  25.800         0   2007.94500         0           0
1337   61  29.070         0  29141.36030         0           1

[1338 rows x 6 columns]

In [81]:
dataset.columns

Index(['age', 'bmi', 'children', 'charges', 'sex_male', 'smoker_yes'], dtype='object')

In [82]:
independent = dataset[['age', 'bmi', 'children', 'sex_male', 'smoker_yes']]

In [83]:
dependent = dataset[['charges']]

In [177]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(independent, dependent, test_size=0.3, random_state=0)

## MultiLinear Regression:

In [12]:
from sklearn.linear_model import LinearRegression
regressor_MultiLinear = LinearRegression()
regressor_MultiLinear.fit(X_train,y_train)

LinearRegression()

In [13]:
regressor_MultiLinear.coef_

array([[  257.8006705 ,   321.06004271,   469.58113407,   -41.74825718,
        23418.6671912 ]])

In [14]:
regressor_MultiLinear.intercept_

array([-12057.244846])

In [15]:
y_pred_MultiLinear = regressor_MultiLinear.predict(X_test)

In [16]:
from sklearn.metrics import r2_score
r_squared_MultiLinear = r2_score(y_pred_MultiLinear,y_test)

In [17]:
r_squared_MultiLinear

0.7100923694262504

## Support Vector Machine Regression:

In [18]:
from sklearn.svm import SVR
regressor_SVR = SVR(kernel="linear")
regressor_SVR.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='linear')

In [19]:
y_pred_SVR = regressor_SVR.predict(X_test)

In [20]:
from sklearn.metrics import r2_score
r_squared_SVR = r2_score(y_pred_SVR,y_test)

In [21]:
r_squared_SVR

-11.396253583524821

### Using scalar values for dependent and independent features to check score improvement is there or not

In [22]:
from sklearn.preprocessing import StandardScaler
standardize = StandardScaler()
scaled_independent = standardize.fit_transform(independent)
scaled_dependent = standardize.fit_transform(dependent)

In [50]:
scaled_independent

array([[-1.43876426, -0.45332   , -0.90861367, -1.0105187 ,  1.97058663],
       [-1.50996545,  0.5096211 , -0.07876719,  0.98959079, -0.5074631 ],
       [-0.79795355,  0.38330685,  1.58092576,  0.98959079, -0.5074631 ],
       ...,
       [-1.50996545,  1.0148781 , -0.90861367, -1.0105187 , -0.5074631 ],
       [-1.29636188, -0.79781341, -0.90861367, -1.0105187 , -0.5074631 ],
       [ 1.55168573, -0.26138796, -0.90861367, -1.0105187 ,  1.97058663]])

In [176]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(scaled_independent,scaled_dependent,test_size=0.3,random_state=0)

In [24]:
from sklearn.svm import SVR
regressor_scaled_SVR = SVR(kernel="linear")
regressor_scaled_SVR.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVR(kernel='linear')

In [25]:
regressor_scaled_SVR.coef_

array([[ 0.30512426,  0.04440183,  0.04301141, -0.01235188,  0.98807753]])

In [26]:
regressor_scaled_SVR.intercept_

array([0.00465516])

In [27]:
regressor_scaled_SVR.n_support_

array([386], dtype=int32)

In [28]:
regressor_scaled_SVR.support_

array([  1,   2,   6,   8,  10,  12,  15,  21,  22,  23,  25,  26,  29,
        32,  33,  34,  35,  36,  37,  39,  41,  45,  46,  48,  52,  58,
        60,  61,  66,  68,  70,  72,  73,  75,  86,  88,  89,  90,  92,
        95,  98, 100, 104, 105, 107, 113, 117, 118, 121, 123, 124, 125,
       126, 127, 130, 138, 139, 143, 149, 153, 158, 160, 167, 173, 177,
       178, 179, 182, 187, 192, 193, 196, 199, 201, 202, 203, 204, 206,
       207, 209, 210, 212, 214, 215, 218, 220, 222, 223, 226, 227, 233,
       235, 236, 238, 239, 240, 242, 243, 247, 248, 251, 257, 259, 263,
       274, 276, 281, 293, 295, 296, 308, 312, 315, 323, 324, 334, 335,
       337, 343, 345, 347, 348, 349, 350, 351, 352, 353, 357, 359, 360,
       361, 363, 364, 365, 366, 369, 370, 373, 375, 376, 377, 379, 382,
       386, 387, 388, 389, 392, 396, 397, 399, 400, 401, 404, 409, 412,
       418, 421, 422, 424, 427, 428, 436, 437, 442, 443, 445, 446, 448,
       449, 450, 453, 456, 458, 459, 460, 462, 468, 470, 474, 47

In [29]:
y_pred_scaled = regressor_scaled_SVR.predict(X_test)

In [30]:
from sklearn.metrics import r2_score
r_squared_scaled = r2_score(y_pred_scaled,y_test)

In [31]:
r_squared_scaled

0.7676100222434411

In [32]:
from sklearn.svm import SVR
regressor_scaled_SVR = SVR(kernel="linear", gamma='scale')
regressor_scaled_SVR.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR1 = SVR(kernel="linear", gamma='auto')
regressor_scaled_SVR1.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR2 = SVR(kernel="poly", gamma='scale')
regressor_scaled_SVR2.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR3 = SVR(kernel="poly", gamma='auto')
regressor_scaled_SVR3.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR4 = SVR(kernel="rbf", gamma='scale')
regressor_scaled_SVR4.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR5 = SVR(kernel="rbf", gamma='auto')
regressor_scaled_SVR5.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR6 = SVR(kernel="sigmoid", gamma='scale')
regressor_scaled_SVR6.fit(X_train,y_train)

from sklearn.svm import SVR
regressor_scaled_SVR7 = SVR(kernel="sigmoid", gamma='auto')
regressor_scaled_SVR7.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1

SVR(gamma='auto', kernel='sigmoid')

In [33]:
y_pred_scaled = regressor_scaled_SVR.predict(X_test)
y_pred_scaled1 = regressor_scaled_SVR1.predict(X_test)
y_pred_scaled2 = regressor_scaled_SVR2.predict(X_test)
y_pred_scaled3 = regressor_scaled_SVR3.predict(X_test)
y_pred_scaled4 = regressor_scaled_SVR4.predict(X_test)
y_pred_scaled5 = regressor_scaled_SVR5.predict(X_test)
y_pred_scaled6 = regressor_scaled_SVR6.predict(X_test)
y_pred_scaled7 = regressor_scaled_SVR7.predict(X_test)

In [34]:
from sklearn.metrics import r2_score
r_squared_scaled = r2_score(y_pred_scaled,y_test)
r_squared_scaled1 = r2_score(y_pred_scaled1,y_test)
r_squared_scaled2 = r2_score(y_pred_scaled2,y_test)
r_squared_scaled3 = r2_score(y_pred_scaled3,y_test)
r_squared_scaled4 = r2_score(y_pred_scaled4,y_test)
r_squared_scaled5 = r2_score(y_pred_scaled5,y_test)
r_squared_scaled6 = r2_score(y_pred_scaled6,y_test)
r_squared_scaled7 = r2_score(y_pred_scaled7,y_test)

In [35]:
print(r_squared_scaled,r_squared_scaled1,r_squared_scaled2,r_squared_scaled3,r_squared_scaled4,r_squared_scaled5,r_squared_scaled6,r_squared_scaled7)

0.7676100222434411 0.7676100222434411 0.853986266611946 0.8539695880515783 0.8693559653084156 0.8693513276175114 0.0015265306824314706 0.028542476831097563


### SVM Regression Conclusion:

### Initially without scaling SVR gives bad accuracy in negative. so, After applying scaling and from all the valuable parameters we got top score for 
### Kernal: rbf for both scale and auto as 0.869 followed by 
### Kernal: poly for both scale and auto as 0.8539

## Decision Tree Regression:

In [36]:
from sklearn.tree import DecisionTreeRegressor

regressor_DTR1 = DecisionTreeRegressor(criterion = 'squared_error', splitter='best')
regressor_DTR2 = DecisionTreeRegressor(criterion = 'squared_error', splitter='random')

regressor_DTR3 = DecisionTreeRegressor(criterion = 'friedman_mse', splitter='best')
regressor_DTR4 = DecisionTreeRegressor(criterion = 'friedman_mse', splitter='random')

regressor_DTR5 = DecisionTreeRegressor(criterion = 'absolute_error', splitter='best')
regressor_DTR6 = DecisionTreeRegressor(criterion = 'absolute_error', splitter='random')

regressor_DTR7 = DecisionTreeRegressor(criterion = 'poisson', splitter='best')
regressor_DTR8 = DecisionTreeRegressor(criterion = 'poisson', splitter='random')

In [37]:
regressor_DTR1.fit(X_train,y_train)
regressor_DTR2.fit(X_train,y_train)
regressor_DTR3.fit(X_train,y_train)
regressor_DTR4.fit(X_train,y_train)
regressor_DTR5.fit(X_train,y_train)
regressor_DTR6.fit(X_train,y_train)

DecisionTreeRegressor(criterion='absolute_error', splitter='random')

In [38]:
y_pred_DTR1 = regressor_DTR1.predict(X_test)
y_pred_DTR2 = regressor_DTR2.predict(X_test)
y_pred_DTR3 = regressor_DTR3.predict(X_test)
y_pred_DTR4 = regressor_DTR4.predict(X_test)
y_pred_DTR5 = regressor_DTR5.predict(X_test)
y_pred_DTR6 = regressor_DTR6.predict(X_test)

In [39]:
from sklearn.metrics import r2_score

r_squared_DTR1 = r2_score(y_pred_DTR1,y_test)
r_squared_DTR2 = r2_score(y_pred_DTR2,y_test)
r_squared_DTR3 = r2_score(y_pred_DTR3,y_test)
r_squared_DTR4 = r2_score(y_pred_DTR4,y_test)
r_squared_DTR5 = r2_score(y_pred_DTR5,y_test)
r_squared_DTR6 = r2_score(y_pred_DTR6,y_test)

In [40]:
print(r_squared_DTR1,r_squared_DTR2,r_squared_DTR3,r_squared_DTR4,r_squared_DTR5,r_squared_DTR6)

0.7512036212690226 0.6837261864046296 0.726991250956752 0.7131677312262047 0.7280456030111537 0.7270773399954908


## Random Forest Regressor:

In [41]:
from sklearn.ensemble import RandomForestRegressor

regressor_RFR1 = RandomForestRegressor(n_estimators=100, criterion= 'squared_error', max_features = 'sqrt')
regressor_RFR2 = RandomForestRegressor(n_estimators=100, criterion= 'squared_error', max_features = 'log2')
regressor_RFR3 = RandomForestRegressor(n_estimators=100, criterion= 'squared_error', max_features = None)

regressor_RFR4 = RandomForestRegressor(n_estimators=100, criterion= 'absolute_error', max_features = 'sqrt')
regressor_RFR5 = RandomForestRegressor(n_estimators=100, criterion= 'absolute_error', max_features = 'log2')
regressor_RFR6 = RandomForestRegressor(n_estimators=100, criterion= 'absolute_error', max_features = None)

regressor_RFR7 = RandomForestRegressor(n_estimators=100, criterion= 'friedman_mse', max_features = 'sqrt')
regressor_RFR8 = RandomForestRegressor(n_estimators=100, criterion= 'friedman_mse', max_features = 'log2')
regressor_RFR9 = RandomForestRegressor(n_estimators=100, criterion= 'friedman_mse', max_features = None)



In [42]:
regressor_RFR1.fit(X_train,y_train)
regressor_RFR2.fit(X_train,y_train)
regressor_RFR3.fit(X_train,y_train)
regressor_RFR4.fit(X_train,y_train)
regressor_RFR5.fit(X_train,y_train)
regressor_RFR6.fit(X_train,y_train)
regressor_RFR7.fit(X_train,y_train)
regressor_RFR8.fit(X_train,y_train)
regressor_RFR9.fit(X_train,y_train)


C:\Anaconda3\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
C:\Anaconda3\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
C:\Anaconda3\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
C:\Anaconda3\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(esti

RandomForestRegressor(criterion='friedman_mse', max_features=None)

In [43]:
y_pred_RFR1 = regressor_RFR1.predict(X_test)
y_pred_RFR2 = regressor_RFR2.predict(X_test)
y_pred_RFR3 = regressor_RFR3.predict(X_test)
y_pred_RFR4 = regressor_RFR4.predict(X_test)
y_pred_RFR5 = regressor_RFR5.predict(X_test)
y_pred_RFR6 = regressor_RFR6.predict(X_test)
y_pred_RFR7 = regressor_RFR7.predict(X_test)
y_pred_RFR8 = regressor_RFR8.predict(X_test)
y_pred_RFR9 = regressor_RFR9.predict(X_test)

In [44]:
from sklearn.metrics import r2_score

r_squared_RFR1 = r2_score(y_pred_RFR1,y_test)
r_squared_RFR2 = r2_score(y_pred_RFR2,y_test)
r_squared_RFR3 = r2_score(y_pred_RFR3,y_test)
r_squared_RFR4 = r2_score(y_pred_RFR4,y_test)
r_squared_RFR5 = r2_score(y_pred_RFR5,y_test)
r_squared_RFR6 = r2_score(y_pred_RFR6,y_test)
r_squared_RFR7 = r2_score(y_pred_RFR7,y_test)
r_squared_RFR8 = r2_score(y_pred_RFR8,y_test)
r_squared_RFR9 = r2_score(y_pred_RFR9,y_test)

In [45]:
print(r_squared_RFR1,r_squared_RFR2,r_squared_RFR3,r_squared_RFR4,r_squared_RFR5,r_squared_RFR6,r_squared_RFR7,r_squared_RFR8,r_squared_RFR9)

0.8472424056220158 0.843610870964661 0.8513917002210595 0.8484342633886321 0.8500878977779167 0.8457614365576245 0.8507438231935716 0.848898759449759 0.8461705728423514


### For Random Forest Algorithm, Parameters ((n_estimators=100, criterion= 'friedman_mse', max_features = None) gives the best score of 0.854

# Regression Assignment Conclusion:

### Multilinear Regression gives the accuracy r_squared score of 0.71

### Support Vector Regression after Scaling gives the maximum accuracy r_squared score of 0.869 for parameters (Kernal: 'rbf' for both gamma: 'scale' and 'auto')

### Decision Tree Regression gives the accuracy r_squared score of 0.75 for parameters (criterion = 'squared_error', splitter='best')

### Random Forest Regression gives the maximum accuracy r_squared score of 0.851 for parameters (n_estimators=100, criterion= 'squared_error', max_features = None)

##  Final Model Selection for Regression Assignment "Insurance" Dataset is SVM with the maximum accuracy r_squared score of 0.869 for parameters (Kernal: 'rbf' for both gamma: 'scale' and 'auto')

In [70]:
import pickle
filename = "FinalModel_RegressionAssignment.sav"
pickle.dump(regressor_scaled_SVR4, open(filename,'wb'))

In [71]:
loaded_model = pickle.load(open("FinalModel_RegressionAssignment.sav",'rb'))
result = loaded_model.predict([[19,27,0,0,1]])

In [73]:
result

array([0.42937651])

## Ada Boost Regressor:

In [85]:
from sklearn.ensemble import AdaBoostRegressor
regressor_ada1 = AdaBoostRegressor(loss = 'linear')
regressor_ada1.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostRegressor()

In [86]:
regressor_ada2 = AdaBoostRegressor(loss = 'square')
regressor_ada2.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostRegressor(loss='square')

In [87]:
regressor_ada3 = AdaBoostRegressor(loss = 'exponential')
regressor_ada3.fit(X_train, y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostRegressor(loss='exponential')

In [88]:
y_pred_ada1 = regressor_ada1.predict(X_test)
y_pred_ada2 = regressor_ada2.predict(X_test)
y_pred_ada3 = regressor_ada3.predict(X_test)

In [89]:
from sklearn.metrics import r2_score
r_squared_ada1 = r2_score(y_pred_ada1,y_test)
r_squared_ada2 = r2_score(y_pred_ada2,y_test)
r_squared_ada3 = r2_score(y_pred_ada3,y_test)

In [90]:
print(r_squared_ada1, r_squared_ada2, r_squared_ada3)

0.8265564488599683 0.28853801700115067 0.5236965975614492


In [153]:
from sklearn.ensemble import AdaBoostRegressor
regressor_ada11 = AdaBoostRegressor(estimator=None,n_estimators=200, learning_rate=2.0, loss = 'linear', random_state=123)

In [154]:
regressor_ada11.fit(X_train,y_train)

C:\Anaconda3\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostRegressor(learning_rate=2.0, n_estimators=200, random_state=123)

In [155]:
y_pred_ada11 = regressor_ada11.predict(X_test)

In [156]:
r_squared_ada11 = r2_score(y_pred_ada11,y_test)

In [157]:
r_squared_ada11

0.852638342020879

### Ada Boost Regression Conclusion:

##### AdaBoostRegressor with parameters loss = 'linear' gives best r_squared score of 0.82655 and Fine tuning of the same parameters gives the best accuracy score of 0.85263 

## XG Boosting Algorithm:

In [167]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import r2_score
regressor_XGBR1 = GradientBoostingRegressor(loss = 'squared_error', n_estimators=100, criterion='friedman_mse')
regressor_XGBR1.fit(X_train,y_train)
y_pred_XGBR1 = regressor_XGBR1.predict(X_test)
r_squared_XGBR1 = r2_score(y_pred_XGBR1,y_test)
print(r_squared_XGBR1)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.8694336572201411


In [168]:
regressor_XGBR2 = GradientBoostingRegressor(loss = 'absolute_error', n_estimators=100, criterion='friedman_mse')
regressor_XGBR2.fit(X_train,y_train)
y_pred_XGBR2 = regressor_XGBR2.predict(X_test)
r_squared_XGBR2 = r2_score(y_pred_XGBR2,y_test)
print(r_squared_XGBR2)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.8017976073090591


In [169]:
regressor_XGBR3 = GradientBoostingRegressor(loss = 'huber', n_estimators=100, criterion='friedman_mse')
regressor_XGBR3.fit(X_train,y_train)
y_pred_XGBR3 = regressor_XGBR3.predict(X_test)
r_squared_XGBR3 = r2_score(y_pred_XGBR3,y_test)
print(r_squared_XGBR3)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.8840438529565386


In [170]:
regressor_XGBR4 = GradientBoostingRegressor(loss = 'quantile', n_estimators=100, criterion='friedman_mse')
regressor_XGBR4.fit(X_train,y_train)
y_pred_XGBR4 = regressor_XGBR4.predict(X_test)
r_squared_XGBR4 = r2_score(y_pred_XGBR4,y_test)
print(r_squared_XGBR4)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.4835729188264162


In [171]:
regressor_XGBR5 = GradientBoostingRegressor(loss = 'squared_error', n_estimators=100, criterion='squared_error')
regressor_XGBR5.fit(X_train,y_train)
y_pred_XGBR5 = regressor_XGBR5.predict(X_test)
r_squared_XGBR5 = r2_score(y_pred_XGBR5,y_test)
print(r_squared_XGBR5)

0.8698451143624354


C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


In [172]:
regressor_XGBR6 = GradientBoostingRegressor(loss = 'absolute_error', n_estimators=100, criterion='squared_error')
regressor_XGBR6.fit(X_train,y_train)
y_pred_XGBR6 = regressor_XGBR6.predict(X_test)
r_squared_XGBR6 = r2_score(y_pred_XGBR6,y_test)
print(r_squared_XGBR6)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.7801551014781289


In [173]:
regressor_XGBR7 = GradientBoostingRegressor(loss = 'huber', n_estimators=100, criterion='squared_error')
regressor_XGBR7.fit(X_train,y_train)
y_pred_XGBR7 = regressor_XGBR7.predict(X_test)
r_squared_XGBR7 = r2_score(y_pred_XGBR7,y_test)
print(r_squared_XGBR7)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.8842399990782791


In [174]:
regressor_XGBR8 = GradientBoostingRegressor(loss = 'quantile', n_estimators=100, criterion='squared_error')
regressor_XGBR8.fit(X_train,y_train)
y_pred_XGBR8 = regressor_XGBR8.predict(X_test)
r_squared_XGBR8 = r2_score(y_pred_XGBR8,y_test)
print(r_squared_XGBR8)

C:\Anaconda3\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


0.4834146969001819


## XGBoost Regressor conclusion:

##### XG Boost Regressor with parameters (loss = 'huber', n_estimators=100, criterion='squared_error') gives the best accuracy score of 0.8842 and
##### parameter (loss = 'huber', n_estimators=100, criterion='squared_error') gives the best accuracy score of 0.8840

## LG Bossting Regression Algorithm ( or LightGradientBoostMachine LGBM Regressor):

In [184]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(independent, dependent, test_size=0.3, random_state=0)

In [185]:
import lightgbm as lgb
regressor_LGBM = lgb.LGBMRegressor(n_estimators=100, learning_rate=0.1, max_depth=-1, random_state=123)
regressor_LGBM.fit(X_train,y_train)

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000401 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 313
[LightGBM] [Info] Number of data points in the train set: 936, number of used features: 5
[LightGBM] [Info] Start training from score 13232.916456


LGBMRegressor(random_state=123)

In [186]:
y_pred_LGBM = regressor_LGBM.predict(X_test)

In [187]:
from sklearn.metrics import r2_score
r_squared_LGBM = r2_score(y_pred_LGBM,y_test)

In [188]:
r_squared_LGBM

0.8592891499824916

##### LGMB Regressor Conclusion:
LGBM Regressor gives the r2_score of 0.859 